### DISCUSSION TITLE
Trigger DataProvider refresh on parent form when modal form closes.

### DISCUSSION THREAD

**Discussion by lukuser** [Budibase/budibase#9745] [Help]
I have a Invoice many-to-one Company structure. (One invoice has one BillTo Company - the same company can be used on many invoices). On the Invoice screen I have a button to pick company, which launches a modal form with a table showing all company records to chose from. User picks a company, which I then write to parent Invoice record using a stored procedure (SQL data source). The reason I use "Execute Query" is because there are several screens using this form to pic a company to be set as FK in their tables (Invoices is just one of them). When the modal form closes, the Invoice form doesn't display the chosen company (the Data Provider doesn't know the FK for BillTO Comapany was updated). Inserting Data Provider refresh action after Navigate To the Company form doesn't work as that executes immediately after the navigate action asynchronously. Is there a way to trigger the parent's DP refresh after the Company pick is made on the modal form?

<sub>[BUDI-6447](https://linear.app/budibase/issue/BUDI-6447/trigger-dataprovider-refresh-on-parent-form-when-modal-form-closes)</sub>

**Comment by aptkingston** [comment]
Hey @lukuser. Data providers will automatically refresh whenever a "write" query occurs to their datasource, and this works when using modals too. If the form on the parent page is using the same data source as the one you write to from the modal then automatic refreshing would work.

Could you clarify what the data source is for both:
- The data provider on the invoice screen
- The "execute query" action

If they're both writing to the same datasource (e.g. the invoices table) then I would expect this to work and I'll investigate further, but if they are using different datasources then automatic refreshing won't work and we'll need to try something else.

Like you said yourself - the refresh data provider action won't work across the modal boundary.

One solution I can suggest, which may or may not work depending on your modal content, would be refactoring that page to use a side panel instead of a modal. Side panels are rendered as normal components, whereas modals are actually a full new client app instance. If you were using side panels, the "refresh data provider" action would work perfectly.

**Comment by lukuser** [comment]
Hello @aptkingston , thank you for your response. Here's the setup of the parent form: 
![image](https://user-images.githubusercontent.com/81827963/216454400-521278ae-ffea-41a2-ac13-e63bec29e07e.png)
 and here's the modal form that appears when I select "Pick from list":
![image](https://user-images.githubusercontent.com/81827963/216456168-31641b55-247a-4e2c-9b18-6a69c9a9da35.png)
The select button on each row would trigger an SP that updates the parent table. URL.type_id is used to determine which parent to update as this form is meant to be used in multiple places.
The actions in "Pick from List" have refresh Data Provider (1st for InvoiceHeader DP, and 2nd for BillToCompanyDP), but that executes before the modal form's query is executed.:
![image](https://user-images.githubusercontent.com/81827963/216457070-8b45889b-9ceb-479e-8cd3-eed577af08c5.png)



The table on the modal for would not fit on a side panel, so I'm still chewing on how I could use it as a picker of companies, where I have to present several fields from the company table and it's related address table for the user to be able to make a choice.

**Comment by aptkingston** [comment]
Another solution you could use then is maybe just not using a modal (uncheck "open screen in modal"), but actually navigating to a new screen to pick a company. When one is picked, you could navigate back to the invoice screen after executing the query. That would also load fresh data.

Just to clarify as well, I can't actually see the data source names for either screen in those screenshots, which would be helpful. Could you show me the data source setting of `BillToCompany_DP` on the parent page and the data source setting that's just cut off screen in your "Execute Query" action? I can see the query name but not the data source.

**Comment by lukuser** [comment]
Hi @aptkingston , here are the two data sources for this part of the form (Step 7). InvoiceHeaderDP's DS in a table Invoice Header .It's linked to its own parent at the top of the form which is the parent for all steps. The linking is done via the filter. The InvoiceHeaderDP's filter is where id (of the InvoiceHeader) = Deal.FK_InvoiceHeader_ID (as represented by {{ Deal_repeater.Deal.id }}). The BillToCompanyDP's filter links to its parent  via id = {{ InvoiceHeaderRepeater.InvoiceHeader.FK_BillToCompany_id }}

![image](https://user-images.githubusercontent.com/81827963/216624270-06e7af21-72db-4fd0-98cb-93a68282b65d.png)
![image](https://user-images.githubusercontent.com/81827963/216624601-37da0853-b196-48a4-9a1f-8106287779d6.png)

As for your idea of turning the modal into a separate screen and navigating back, I'd like to do that as long as I can somehow navigate back to the correct step of the parent form  (Step 7 - I'm using the stepform control). Is there a way to specify that on the navigation action or any other way?

**Comment by NicolasFabre81** [comment]
I have a similar issue (V2.5) : refresh does not work whereas I use the same SQL table in both parent and modal screens.
My modal operates a SQL update, that should make a row appear in the parent screen table.
The row appears only if I click my refresh button.

Is it supposed to work on UPDATEs ? Or maybe only on INSERTs ?

**Reply by gbarte** [reply]
I have this issue too, the data provider doesn't refresh after I UPDATE data from a modal
